Panic in thirdpartyresourcedata/codec.go

<!-- Thanks for filing an issue! Before hitting the button, please answer these questions.-->

**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No


**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): panic thirdpartyresourcedata

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

<!--
If this is a BUG REPORT, please:
  - Fill in as much of the template below as you can.  If you leave out
    information, we can't help you as well.

If this is a FEATURE REQUEST, please:
  - Describe *in detail* the feature/behavior/change you'd like to see.

In both cases, be ready for followup questions, and please respond in a timely
manner.  If we can't reproduce a bug or think a feature already exists, we
might close your issue.  If we're wrong, PLEASE feel free to reopen it and
explain why.
-->

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:16:57Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0+coreos.0", GitCommit:"278a1f7034bdba61cba443722647da1a8204a6fc", GitTreeState:"clean", BuildDate:"2016-09-26T20:48:37Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```
Also present in current master

**What happened**:
As part of attempt to support third party resources in kubernetes/helm#1295, I've found that the default client attempts a weird patch behaviour:
```
PATCH /apis/stable.k8s.psg.io/v1/namespaces/hello-world/certificates/hello-world HTTP/1.1
Host: localhost:8080
User-Agent: tiller/v1.4.1 (linux/amd64) kubernetes/$Format
Content-Length: 388
Accept: application/json, */*
Content-Type: application/strategic-merge-patch+json
Accept-Encoding: gzip

{"data":"eyJhcGlWZXJzaW9uIjoic3RhYmxlLms4cy5wc2cuaW8vdjEiLCJraW5kIjoiQ2VydGlmaWNhdGUiLCJtZXRhZGF0YSI6eyJuYW1lIjoiaGVsbG8td29ybGQifSwic3BlYyI6eyJkb21haW4iOiJoZWxsby13b3JsZC5rOHMuZXUtd2VzdC0xLmF3cy5zaGQuZGV2Lmxhc3RtaWxlLmNvbSIsImVtYWlsIjoibStoZWxsby13b3JsZEBvY2Fkby5jb20iLCJwcm92aWRlciI6InJvdXRlNTMifX0=","kind":"ThirdPartyResourceData","metadata":{"labels":null},"spec":null,"status":null}
```

This causes kube-apiserver to panic:
```
I1010 16:04:50.014030       1 handlers.go:162] GET /apis/stable.k8s.psg.io/v1/namespaces/hello-world/certificates/hello-world: (1.486244ms) 200 [[tiller/v1.4.1 (linux/amd64) kubernetes/$Format] 127.0.0.1:53188]
E1010 16:04:50.033872       1 runtime.go:64] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:70
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:63
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:49
/usr/local/go/src/runtime/asm_amd64.s:472
/usr/local/go/src/runtime/panic.go:443
/usr/local/go/src/runtime/panic.go:62
/usr/local/go/src/runtime/sigpanic_unix.go:24
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/registry/thirdpartyresourcedata/codec.go:421
<autogenerated>:5
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/runtime/codec.go:60
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:573
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/api/rest/update.go:177
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/registry/generic/registry/store.go:355
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go:462
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/registry/generic/registry/store.go:431
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:645
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:953
/usr/local/go/src/runtime/asm_amd64.s:1998
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x0 pc=0x15e2192]

goroutine 1212 [running]:
panic(0x4475c00, 0xc820016070)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
k8s.io/kubernetes/pkg/util/runtime.HandleCrash(0xc8223d3ef8, 0x1, 0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:56 +0x153
panic(0x4475c00, 0xc820016070)
	/usr/local/go/src/runtime/panic.go:443 +0x4e9
k8s.io/kubernetes/pkg/registry/thirdpartyresourcedata.(*thirdPartyResourceDataDecoder).Decode(0xc821509740, 0xc821735880, 0x223, 0x35c, 0x0, 0x7fbdf5c06130, 0xc822bdc480, 0x0, 0x0, 0x10739d9, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/registry/thirdpartyresourcedata/codec.go:421 +0x1dc2
k8s.io/kubernetes/pkg/runtime.(*codec).Decode(0xc821509760, 0xc821735880, 0x223, 0x35c, 0x0, 0x7fbdf5c06130, 0xc822bdc480, 0x0, 0x0, 0xc8226a2900, ...)
	<autogenerated>:5 +0xca
k8s.io/kubernetes/pkg/runtime.DecodeInto(0x7fbdf5bc4f88, 0xc821509760, 0xc821735880, 0x223, 0x35c, 0x7fbdf5c06130, 0xc822bdc480, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/runtime/codec.go:60 +0x98
k8s.io/kubernetes/pkg/apiserver.patchResource.func1(0x7fbdf5c46f90, 0xc822ba05d0, 0x0, 0x0, 0x7fbdf5c06130, 0xc822ed5680, 0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:573 +0x11f7
k8s.io/kubernetes/pkg/api/rest.(*defaultUpdatedObjectInfo).UpdatedObject(0xc822ba6240, 0x7fbdf5c46f90, 0xc822ba05d0, 0x7fbdf5c06130, 0xc822ed5680, 0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/api/rest/update.go:177 +0x1aa
k8s.io/kubernetes/pkg/registry/generic/registry.(*Store).Update.func1(0x7fbdf5c06130, 0xc822ed5680, 0x0, 0x57846d, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/registry/generic/registry/store.go:355 +0x10d
k8s.io/kubernetes/pkg/storage/etcd.(*etcdHelper).GuaranteedUpdate(0xc8201cd700, 0x7fbdf5c46f50, 0xc822ba05d0, 0xc822ba2ae0, 0x57, 0x7fbdf5c06130, 0xc822ed5560, 0x1, 0xc820028af8, 0xc822bcc000, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go:462 +0x49e
k8s.io/kubernetes/pkg/registry/generic/registry.(*Store).Update(0xc82018b5e0, 0x7fbdf5c46f90, 0xc822ba05d0, 0xc822ba2645, 0xb, 0x7fbdf5bd9b70, 0xc822ba6240, 0x0, 0x0, 0x40a100, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/registry/generic/registry/store.go:431 +0x401
k8s.io/kubernetes/pkg/apiserver.patchResource.func3(0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:645 +0xda
k8s.io/kubernetes/pkg/apiserver.finishRequest.func1(0xc822ba2900, 0xc822ea88c0, 0xc822ba28a0, 0xc822ba2840)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:953 +0xd9
created by k8s.io/kubernetes/pkg/apiserver.finishRequest
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/apiserver/resthandler.go:958 +0xf1
m@hotcpc17686:/export/home/m/go/src/k8s.io/kubernetes$

```

**What you expected to happen**:
No panic. I suspect this behaviour shouldn't work anyway, but it should return an error rather than letting a user crash the apiserver at-will.

**How to reproduce it** (as minimally and precisely as possible):
* Use a third party resource such as the Certificate one from https://github.com/PalmStoneGames/kube-cert-manager
* Create a resource
* Patch it with something like the following: 
```
{"data":"eyJhcGlWZXJzaW9uIjoic3RhYmxlLms4cy5wc2cuaW8vdjEiLCJraW5kIjoiQ2VydGlmaWNhdGUiLCJtZXRhZGF0YSI6eyJuYW1lIjoiaGVsbG8td29ybGQifSwic3BlYyI6eyJkb21haW4iOiJoZWxsby13b3JsZC5rOHMuZXUtd2VzdC0xLmF3cy5zaGQuZGV2Lmxhc3RtaWxlLmNvbSIsImVtYWlsIjoibStoZWxsby13b3JsZEBvY2Fkby5jb20iLCJwcm92aWRlciI6InJvdXRlNTMifX0="
,"kind":"ThirdPartyResourceData"}
```

**Anything else do we need to know**:
PR to follow with a test case, no fix yet

--- a/pkg/registry/extensions/thirdpartyresourcedata/codec.go
+++ b/pkg/registry/extensions/thirdpartyresourcedata/codec.go
@@ -418,7 +418,7 @@ func (t *thirdPartyResourceDataDecoder) Decode(data []byte, gvk *unversioned.Gro
 			return nil, nil, fmt.Errorf("unexpected object for 'kind': %v", kindObj)
 		}
 		if len(t.kind) > 0 && kindStr != t.kind {
-			return nil, nil, fmt.Errorf("kind doesn't match, expecting: %s, got %s", gvk.Kind, kindStr)
+			return nil, nil, fmt.Errorf("kind doesn't match, expecting: %s, got %s", t.kind, kindStr)
 		}
 		actual.Kind = kindStr
 	}

--- a/pkg/registry/extensions/thirdpartyresourcedata/codec_test.go
+++ b/pkg/registry/extensions/thirdpartyresourcedata/codec_test.go
@@ -81,6 +81,15 @@ func TestCodec(t *testing.T) {
 			name: "basic",
 		},
 		{
+			into: &extensions.ThirdPartyResourceData{},
+			obj: &Foo{
+				ObjectMeta: api.ObjectMeta{Name: "bar"},
+				TypeMeta:   unversioned.TypeMeta{Kind: "ThirdPartyResourceData"},
+			},
+			expectErr: true,
+			name:      "broken kind",
+		},
+		{
 			obj: &Foo{
 				ObjectMeta: api.ObjectMeta{Name: "bar", ResourceVersion: "baz"},
 				TypeMeta:   unversioned.TypeMeta{APIVersion: "company.com/v1", Kind: "Foo"},

